DOCS-8072-4.3-tuning-logging-kt (#1420)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-logging.adoc

> = Logging

Reviewer: ```suggestion
= Asynchronous vs. Synchronous Logging
```

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Mule runtime engine (Mule) 4 uses asynchronous logging by default, so the logging operation happens in a separate thread from the message processing, and the latter can continue without having to wait for the log operation to complete. This occurs by using Log4j 2 as default logging.

Reviewer: Shorten and simplify
```suggestion
Mule runtime engine (Mule) 4 uses Log4j 2 asynchronous logging by default, which enables message processing to continue while logging occurs in a separate thread. +
However, because asynchronous logging poses a performance-reliability trade-off (https://blogs.mulesoft.com/dev/mule-dev/mule-3-6-asynchronous-logging/[Asynchronous Logging Mule]), best practice is to use asynchronous logging with a minimum logging level of `WARN` for a production application. +
Use asynchronous logging with a minimum logging level of  `INFO` when you need to confirm events such as successful policy installation or to perform troubleshooting.
```

As the thread processing your message does not have to wait for the log message to be fully handled before it can continue, you can achieve better performance than in synchronous logging strategy. +

> Alternatively, the use of asynchronous logging may pose a performance-reliability trade-off as explained in https://blogs.mulesoft.com/dev/mule-dev/mule-3-6-asynchronous-logging/[Asynchronous Logging Mule], and you may lose some messages if Mule crashes before the logging buffers are flushed to the disk. In this case, keep in mind that you can have a mixed configuration of asynchronous/synchronous loggers in your app.

Reviewer: replaced by suggestion above

> The general recommendation is to favor asynchronous logging over synchronous one and to set logging level `WARN` as the minimum for a production application. However, in some special cases, you might want to enable `INFO` logging level for specific packages to spot certain events such as policies successful installation and for troubleshooting purposes.

Reviewer: ditto

> You can configure the logging strategy by editing the `src/main/resources/log4j2.xml` file of your application. The following shows a sample configuration file:

Reviewer: ```suggestion
Configure your logging strategy by editing your application's `src/main/resources/log4j2.xml` file, as shown in the following example:
```

[source,xml,linenums]
----
<?xml version="1.0" encoding="utf-8"?>
<Configuration>
   <Appenders>
       <RollingFile name="file" fileName="${sys:mule.home}${sys:file.separator}logs${sys:file.separator}try-transactional-propagate.log"
                filePattern="${sys:mule.home}${sys:file.separator}logs${sys:file.separator}try-transactional-propagate-%i.log">
           <PatternLayout pattern="%-5p %d [%t] [event: %X{correlationId}] %c: %m%n" />
           <SizeBasedTriggeringPolicy size="10 MB" />
           <DefaultRolloverStrategy max="10"/>
       </RollingFile>
   </Appenders>
   <Loggers>

       <!-- Http Logger shows wire traffic on DEBUG. -->
       <!--AsyncLogger name="org.mule.service.http.impl.service.HttpMessageLogger" level="DEBUG" /-->
       <AsyncLogger name="org.mule.service.http" level="WARN"/>
       <AsyncLogger name="org.mule.extension.http" level="WARN"/>

   <!-- Mule logger -->
       <AsyncLogger name="org.mule.runtime.core.internal.processor.LoggerMessageProcessor" level="WARN"/>
       <AsyncRoot level="WARN">
           <AppenderRef ref="file" />
       </AsyncRoot>
   </Loggers>
</Configuration>
----

> Apply caution when logging parts of messages, especially the payload. Logging messages will bypass streaming and incur significant disk IO for large payloads.

Reviewer: ```suggestion
Apply caution when logging parts of messages, especially the payload, because this bypasses streaming and incurs significant disk I/O.
```
I wonder if this warning should be made before the example?
Is message logging the primary type of logging, or one common type, or an infrequent type?
